weather_mv/loader_pipeline/metrics.py

```diff
@@ -170,8 +178,90 @@ def process(self, element):
                 logger.info(
                     f"{uri}: Time from {current_step} -> {next_step}: {step_time} seconds."
                 )
-
+            yield ("custom_metrics", (data_latency_ms / 1000, element_processing_time / 1000))
         except Exception as e:
             logger.warning(
                 f"Some error occured while adding metrics. Error {e}"
             )
+
+
+@dataclasses.dataclass
+class CreateTimeSeries(beam.DoFn):
+    """DoFn to write metrics TimeSeries data in Google Cloud Monitoring."""
+    job_name: str
+    project: str
+    region: str
+
+    def create_time_series(self, metric_name: str, metric_value: float) -> None:
+        """Writes data to a Metrics TimeSeries."""
+        client = monitoring_v3.MetricServiceClient()
+        series = monitoring_v3.TimeSeries()
+        series.metric.type = f"custom.googleapis.com/{metric_name}"
+        series.metric.labels["description"] = metric_name
+        series.resource.type = "dataflow_job"
+        series.resource.labels["job_name"] = self.job_name
+        series.resource.labels["project_id"] = self.project
+        series.resource.labels["region"] = self.region
+
+        now = time.time()
+        seconds = int(now)
+        nanos = int((now - seconds) * 10**9)
+        interval = monitoring_v3.TimeInterval(
+            {"end_time": {"seconds": seconds, "nanos": nanos}}
+        )
+
+        point = monitoring_v3.Point(
+            {"interval": interval, "value": {"double_value": metric_value}}
+        )
+        series.points = [point]
+        client.create_time_series(
+            name=f"projects/{self.project}", time_series=[series]
+        )
+        logger.info(
+            f"Successfully created time series for {metric_name}. Metric value: {metric_value}."
+        )
+
+    def process(self, element: t.Any):
+        _, metric_values = element
+        data_latency_times = [x[0] for x in metric_values]
+        element_processing_times = [x[1] for x in metric_values]
+
+        logger.info(f"data_latency_time values: {data_latency_times}")
+        self.create_time_series("data_latency_time_max", max(data_latency_times))
+        self.create_time_series(
+            "data_latency_time_mean", sum(data_latency_times) / len(data_latency_times)
+        )
+
+        logger.info(f"element_processing_time values: {element_processing_times}")
+        self.create_time_series("element_processing_time_max", max(element_processing_times))
+        self.create_time_series(
+            "element_processing_time_mean", sum(element_processing_times) / len(element_processing_times)
+        )
```

Comment:
Combining 4 separate requests in 1 should be efficient. I believe we can write points in different timeseries using a single request.

Reply:
Acknowledged. Changes done.